csrc/auto_free.h

```diff
@@ -0,0 +1,183 @@
+    public:                                             \
+        AUTO_CONSTRUCTORS(name)                         \
+        CLASSNAME(name)()                               \
+        {                                               \
+            PTR_NAME(name) = NULL;                      \
+        }                                               \
+        static CLASSNAME(name) from(name *ptr)          \
+        {                                               \
+            CLASSNAME(name) tmp;                        \
+            tmp.PTR_NAME(name) = ptr;                   \
+            return tmp;                                 \
+        }                                               \
+        ~CLASSNAME(name)()                              \
+        {                                               \
+            clear();                                    \
+        }                                               \
+        bool isInitialized()                            \
+        {                                               \
+            return !!PTR_NAME(name);                    \
+        }                                               \
+        bool set(name *ptr)                             \
+        {                                               \
+            clear();                                    \
+            PTR_NAME(name) = ptr;                       \
+            return !!ptr;                               \
+        }                                               \
+        name *take()                                    \
+        {                                               \
+            name *tmpPtr = PTR_NAME(name);              \
+            PTR_NAME(name) = NULL;                      \
+            return tmpPtr;                              \
+        }                                               \
+        void releaseOwnership()                         \
+        {                                               \
+            PTR_NAME(name) = NULL;                      \
+        }                                               \
+        void clear()                                    \
+        {                                               \
+            CONCAT2(name, _free)(PTR_NAME(name));       \
+            PTR_NAME(name) = NULL;                      \
+        }                                               \
+        name *operator->()                              \
+        {                                               \
+            return *this;                               \
+        }                                               \
+        operator name *()                               \
+        {                                               \
+            if (!PTR_NAME(name))                        \
+            {                                           \
+                abort();                                \
```

Comment:
do we handle this SIGABRT anywhere? or is the intention to just terminate the program if this case is hit?